h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/util/NativeLibrary.java

```diff
@@ -112,27 +107,52 @@ private ClassLoader getClassLoader() {
     return classLoader;
   }
 
-  private void extractAndLoad(String ...libPaths) throws IOException {
-    Throwable lastException = null;
-    for (String libPath : libPaths) {
-      try {
-        lastException = null;
-        File temp = extract(libPath, getClassLoader());
-        // Finally, load the library
-        System.load(temp.getAbsolutePath());
-        // Perfect loaded, break the cycle
-        logger.info("Loaded library from " + libPath + " (" + temp.getAbsolutePath() + ")");
-        break;
-      } catch (IOException | UnsatisfiedLinkError e) {
-        logger.warn("Cannot load library from path " + libPath);
-        lastException = e;
+  private boolean extractAndLoad(String libPath) throws IOException {
+    try {
+      URL libResource = getLibResource(libPath, getClassLoader());
+      if (libResource == null) {
+        logger.debug("We don't bundle library " + libPath);
+        return false;
       }
+      File temp = extract(libPath, libResource);
+      // Finally, load the library
+      System.load(temp.getAbsolutePath());
+      // Perfect loaded, break the cycle
+      logger.info("Loaded library from " + libPath + " (" + temp.getAbsolutePath() + ")");
+      return true;
+    } catch (IOException | UnsatisfiedLinkError e) {
+      logger.warn("Cannot load library from path " + libPath);
+      throw new IOException(e);
     }
-    if (lastException != null) throw new IOException(lastException);
   }
 
-  private static File extract(String libPath, ClassLoader classLoader)
+  private URL getLibResource() {
+    return getLibResource(getPlatformLibraryPath(), getClassLoader());
+  }
+
+  public boolean isBundled() {
+    return getLibResource() != null;
+  }
+
+  public File extractTo(File directory) throws IOException {
+    File target = new File(directory, platform.getPlatformLibName(getName()));
+    extractTo(getLibResource(), target);
+    return target;
+  }
+  
+  private static void extractTo(URL libResource, File target) throws IOException {
+    try (InputStream is = libResource.openStream()) {
+      Files.copy(is, target.toPath(), StandardCopyOption.REPLACE_EXISTING);
+    }
+  }
+
+  private static URL getLibResource(String libPath, ClassLoader classLoader) {
+    return classLoader.getResource(libPath);
+  }
+
+  private static File extract(String libPath, URL libResource)
       throws IOException, IllegalArgumentException {
+    assert libResource != null;
```

Comment:
Some message about why it should not be null, what we expect here?

Reply:
fixed, thank you!